chainerrl/agents/iqn.py

```diff
@@ -0,0 +1,355 @@
+        """Evaluate given observations.
+
+        Args:
+            x (ndarray): Batch of observations.
+        Returns:
+            callable: (batch_size, taus) -> (batch_size, taus, n_actions)
+        """
+        batch_size = x.shape[0]
+        psi_x = self.psi(x)
+        assert psi_x.ndim == 2
+        assert psi_x.shape[0] == batch_size
+        hidden_size = psi_x.shape[1]
+
+        def evaluate_with_quantile_thresholds(taus):
+            assert taus.ndim == 2
+            assert taus.shape[0] == batch_size
+            n_taus = taus.shape[1]
+            phi_taus = self.phi(taus)
+            assert phi_taus.ndim == 3
+            assert phi_taus.shape == (batch_size, n_taus, hidden_size)
+            psi_x_b = F.broadcast_to(
+                F.expand_dims(psi_x, axis=1), phi_taus.shape)
+            h = psi_x_b * phi_taus
+            h = F.reshape(h, (-1, hidden_size))
+            assert h.shape == (batch_size * n_taus, hidden_size)
+            h = self.f(h)
+            assert h.ndim == 2
+            assert h.shape[0] == batch_size * n_taus
+            n_actions = h.shape[-1]
+            h = F.reshape(h, (batch_size, n_taus, n_actions))
+            return QuantileDiscreteActionValue(h)
+
+        return evaluate_with_quantile_thresholds
+
+
+def _unwrap_variable(x):
+    if isinstance(x, chainer.Variable):
+        return x.array
+    else:
+        return x
+
+
+def compute_eltwise_huber_quantile_loss(y, t, taus, huber_loss_threshold=1.0):
+    """Compute elementwise Huber losses for quantile regression.
+
+    Args:
+        y (chainer.Variable): (batch_size, N)
+        t (chainer.Variable or ndarray): (batch_size, N_prime)
+        taus (ndarray): (batch_size, N)
+        huber_loss_threshold (float): Threshold of Huber loss. In the IQN
+            paper, this is denoted by kappa.
+
+    Returns:
+        chainer.Variable: Loss (batch_size, N, N_prime)
+    """
+    assert y.shape == taus.shape
+    # (batch_size, N) -> (batch_size, N, 1)
+    y = F.expand_dims(y, axis=2)
+    # (batch_size, N_prime) -> (batch_size, 1, N_prime)
+    t = F.expand_dims(t, axis=1)
+    # (batch_size, N) -> (batch_size, N, 1)
+    taus = F.expand_dims(taus, axis=2)
+    # Broadcast to (batch_size, N, N_prime)
+    y, t, taus = F.broadcast(y, t, taus)
+    with chainer.no_backprop_mode():
+        I_delta = ((_unwrap_variable(t) - _unwrap_variable(y)) > 0).astype('f')
+    eltwise_huber_loss = F.huber_loss(
+        y, t, delta=huber_loss_threshold, reduce='no')
+    eltwise_loss = abs(taus - I_delta) * eltwise_huber_loss
+    return eltwise_loss
+
+
+class IQN(dqn.DQN):
+    """Implicit Quantile Networks.
+
+    See https://arxiv.org/abs/1806.06923.
+
+    Args:
+        quantile_thresholds_N (int): Number of quantile thresholds used in
+            quantile regression.
+        quantile_thresholds_N_prime (int): Number of quantile thresholds used
+            to sample from the return distribution at the next state.
```

Comment:
These two args are about numbers of samples.  The name "thresholds" confused me.